### PR TITLE
[SPARK-35669][SQL] Quote the pushed column name only when nested column predicate pushdown is enabled

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/31964

We should only quote the column name when nested column predicate pushdown is enabled, otherwise the data source side may not have the logic to parse the quoted column name and fail. This is not a problem before #31964 , as we don't quote the column name if there is no dot in the name. But #31964 changed it.

### Why are the changes needed?

fix a query failure

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sarutak @tanelk

**Comment by HyukjinKwon** [comment]
cc @MaxGekk too

**Comment by HyukjinKwon** [comment]
Shall we revert https://github.com/apache/spark/pull/31964?

**Comment by sarutak** [comment]
#31964 is mostly for the display format of queries so we can revert it safely. But even if we revert it, the potential problem is present isn't it?
This problem happens with ```col("`a``b`")``` even before #31964.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -699,20 +699,25 @@ abstract class PushableColumnBase {
...
-        // nested predicate pushdown is enabled.
-        if (nestedPredicatePushdownEnabled || !a.name.contains(".")) {
-          Some(Seq(a.name))
-        } else {
-          None
-        }
-      case s: GetStructField if nestedPredicatePushdownEnabled =>
-        helper(s.child).map(_ :+ s.childSchema(s.ordinal).name)
-      case _ => None
+    if (nestedPredicatePushdownEnabled) {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:702]: cc @dbtsai @viirya

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala [hunk 2]
```diff
@@ -35,8 +35,9 @@ sealed abstract class Filter {
   /**
    * List of columns that are referenced by this filter.
    *
-   * Note that, each element in `references` represents a column; `dots` are used as separators
-   * for nested columns. If any part of the names contains `dots`, it is quoted to avoid confusion.
+   * Note that, each element in `references` represents a column. The column name follows ANSI SQL
+   * names and identifiers: dots are used as separators for nested columns, name will be quoted if
+   * it contains special chars.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala:40]: So it extends the coverage of quoted name (originally it is only for non-separator dots)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:702]: looks good.

[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:702]: Only exceptional one might be, if a v2 implementation can only handle quoted "dots", but not other quoted chars, right? I think it should be rare cases.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala:40]: yea

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:702]: Yea I think that's rare. V2 implementations should already be using a decent parser that recognizes dots and backticks.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
cc @maropu for another look.

**Comment by cloud-fan** [comment]
ping @maropu

**Comment by viirya** [comment]
This looks good. I'll merge this later if no more comments.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 3]
```diff
@@ -699,20 +699,25 @@ abstract class PushableColumnBase {
...
+  }
+
+  private def extractNameBasic(e: Expression): Option[String] = e match {
+    // Attribute that contains dot "." in name is supported only when nested predicate pushdown
+    // is enabled.
+    case a: Attribute if !a.name.contains(".") => Some(a.name)
+    case _ => None
+  }
+
+  private def extractNameAdvanced(e: Expression): Option[Seq[String]] = e match {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:716]:
```suggestion
  private def extractQualifiedName(e: Expression): Option[Seq[String]] = e match {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 4]
```diff
@@ -699,20 +699,25 @@ abstract class PushableColumnBase {
...
-      case _ => None
+    if (nestedPredicatePushdownEnabled) {
+      extractNameAdvanced(e).map(_.quoted)
+    } else {
+      extractNameBasic(e)
     }
-    helper(e).map(_.quoted)
+  }
+
+  private def extractNameBasic(e: Expression): Option[String] = e match {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:709]:
```suggestion
  private def extractPlainName(e: Expression): Option[String] = e match {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 5]
```diff
@@ -699,20 +699,25 @@ abstract class PushableColumnBase {
...
-        } else {
-          None
-        }
-      case s: GetStructField if nestedPredicatePushdownEnabled =>
-        helper(s.child).map(_ :+ s.childSchema(s.ordinal).name)
-      case _ => None
+    if (nestedPredicatePushdownEnabled) {
+      extractNameAdvanced(e).map(_.quoted)
+    } else {
+      extractNameBasic(e)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:705]:
```suggestion
      extractPlainName(e)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 6]
```diff
@@ -699,20 +699,25 @@ abstract class PushableColumnBase {
...
-        if (nestedPredicatePushdownEnabled || !a.name.contains(".")) {
-          Some(Seq(a.name))
-        } else {
-          None
-        }
-      case s: GetStructField if nestedPredicatePushdownEnabled =>
-        helper(s.child).map(_ :+ s.childSchema(s.ordinal).name)
-      case _ => None
+    if (nestedPredicatePushdownEnabled) {
+      extractNameAdvanced(e).map(_.quoted)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:703]:
```suggestion
      extractQualifiedName(e).map(_.quoted)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 7]
```diff
@@ -699,20 +699,25 @@ abstract class PushableColumnBase {
...
+    // Attribute that contains dot "." in name is supported only when nested predicate pushdown
+    // is enabled.
+    case a: Attribute if !a.name.contains(".") => Some(a.name)
+    case _ => None
+  }
+
+  private def extractNameAdvanced(e: Expression): Option[Seq[String]] = e match {
+    case a: Attribute => Some(Seq(a.name))
+    case s: GetStructField =>
+      extractNameAdvanced(s.child).map(_ :+ s.childSchema(s.ordinal).name)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala:719]:
```suggestion
      extractQualifiedName(s.child).map(_ :+ s.childSchema(s.ordinal).name)
```

**Comment by HyukjinKwon** [comment]
@cloud-fan, would you mind fixing the PR title to reflect the change itself instead of the issue being fixed?

**Review by maropu — APPROVED** [review]
Looks fine.

**Comment by maropu** [comment]
Thanks, merged to master.

**Pull request closed without merging**